List more math function in API docs

brandonhorsley opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-devs/pymc/releases -->

## Description
<!--- Describe your changes in detail -->
This commit is a solution to the problems presented in PyMC github issue #7130. The first part of the issue was that certain functions like inverse trigonometric functions didn't seem to work, however on a clean install for this commit and pull request, that issue disappeared and the functions work as intended. These functions were missing among others from the website documentation which is the second part of the github issue and is what this particular git commit concerns. Note I have included tround although other codes reflect this as being deprecated but still able to be used. I have also rearranged the ordering to better reflect similar functions being bunched together.
## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [x] Closes #7130
- [] Related to # 

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [ ] New feature / enhancement
- [ ] Bug fix
- [x] Documentation
- [ ] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->


<!-- readthedocs-preview pymc start -->
----
📚 Documentation preview 📚: https://pymc--7211.org.readthedocs.build/en/7211/

<!-- readthedocs-preview pymc end -->

brandonhorsley committed:
This commit is a partial solution to the problems presented in PyMC github issue #7130. The first part of the issue was that certain functions like inverse trigonometric functions didn't seem to work, however on a clean install for this commit and pull request, that issue disappeared and the functions work as intended. These functions were missing among others from the website documentation which is the second part of the github issue and is what this particular git commit concerns. Note I have included tround although other codes reflect this as being deprecated but still able to be used. I have also rearranged the ordering to better reflect similar functions being bunched together.

OriolAbril reviewed (approved):
Everything looks good, I agree with sorting them into multiple topic related subsections but can be done in a follow up PR

brandonhorsley committed:
Removed numpy functions and tround from math.rst

brandonhorsley committed:
Edits to math.py.

Removed tround, replaced round with pytensor.tensor import, and added FutureWarning to log1mexp_numpy and logdiffexp_numpy.

ricardoV94 commented on pymc/math.py:
```diff
@@ -272,27 +274,18 @@ def kron_diag(*diags):
 def logdiffexp_numpy(a, b):
     """log(exp(a) - exp(b))"""
+    warnings.warn(
+        "pymc.math.logdiffexp_numpy is being deprecated. Use logdiffexp instead.",
```
They are not equivalent. We don't need to offer an alternative

```suggestion
        "pymc.math.logdiffexp_numpy is being deprecated.",
```

ricardoV94 commented on pymc/math.py:
```diff
@@ -341,6 +334,11 @@ def log1mexp_numpy(x, *, negative_input=False):
     https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf
     """
+    warnings.warn(
+        "pymc.math.log1mexp_numpy is being deprecated. Use log1mexp instead.",
```
```suggestion
        "pymc.math.log1mexp_numpy is being deprecated.",
```

ricardoV94 replied:
Also need to implement it locally here, so that it can be safely removed later:

https://github.com/pymc-devs/pymc/blob/f302882da263d8569af1f55b3d758c614ecb3799/tests/distributions/test_continuous.py#L427

brandonhorsley replied:
OK, will do, I'll check for all cases where those numpy functions show up, is that just to have the comment visually written there so it can be identified from looking at the file?

ricardoV94 replied:
Sorry, what do you mean?

brandonhorsley replied:
As in since the warning is applied to the function definition 'log1mexp_numpy' the warning would be raised when those codes are run, so if I'm understanding correctly I am to write another warning on that test_continuous.py line so when you talk about safely removing it, I assume that means that by writing the warning code there that you know that it is to be removed/altered there as well by just looking at the line of code itself without needing to run the script to get the warning If that makes sense?

I also see there is another file where the function comes up as well so I will edit that too and I will do a search for cases where logdiffexp_numpy pops up and another search in case tround happens to show up anywhere else too

ricardoV94 replied:
I mean no other part of the codebase should be using a function marked for deprecation (other than code that is testing that function directly). That ensures then when the time comes all we need to do is remove the deprecated code.

brandonhorsley committed:
Update pymc/math.py

Co-authored-by: Ricardo Vieira <28983449+ricardoV94@users.noreply.github.com>

brandonhorsley committed:
Update pymc/math.py

Co-authored-by: Ricardo Vieira <28983449+ricardoV94@users.noreply.github.com>

brandonhorsley committed:
Further FutureWarnings applied to _numpy functions

Searched PyMC codebase for occurrences of _numpy functions and added
explicit warnings. Most of these were for log1mexp_numpy which had quite
a few involvements in test codes. Also note that on test_math.py the
_numpy functions are explicitly imported and so when deprecated those
imports will have to be removed too.

ricardoV94 commented on tests/distributions/test_continuous.py:
```diff
@@ -418,6 +418,7 @@ def scipy_log_pdf(value, a, b):
 
         def scipy_log_cdf(value, a, b):
+            warnings.warn("pymc.math.log1mexp_numpy is being deprecated.", FutureWarning)
             return pm.math.log1mexp_numpy(b * np.log1p(-(value**a)), negative_input=True)
```
The test function doesn't need to issue a warning. Instead it should not use `math.log1mex_numpy` but a non user facing helper that achieves the same

brandonhorsley replied:
Ahh OK

ricardoV94 commented on tests/test_math.py:
```diff
@@ -148,6 +148,8 @@ def test_log1mexp():
         warnings.filterwarnings("ignore", "divide by zero encountered in log", RuntimeWarning)
         warnings.filterwarnings("ignore", "invalid value encountered in log", RuntimeWarning)
+
+        warnings.warn("pymc.math.log1mexp_numpy is being deprecated.", FutureWarning)
```
No need to issue the warning in the tests. Test code is not user facing. Same for all other warnings below.

brandonhorsley replied:
OK so I can take out those warnings and I presume then it is the same as the test_continuous case of just substituting in alternative code that achieves the same or is it that i should just undo my edits in test_math.py since they are test codes?

ricardoV94 replied:
This is different from the `test_continuous`. This is testing the functions directly, so can stay as is. This code will be removed when the respective functions are removed.

The `test_continuous` is the only one that needs a separate version of the function. As it will still need it when we remove the deprecated code.

brandonhorsley committed:
Undid test_math edits, test_continuous change

This commit undoes the changes to test_math.py by removing the
FutureWarning warnings for the _numpy functions. Note that when these
tests are removed the _numpy functions are imported explicitly in this
file. Replaced the log1mexp_numpy use in test_continuous.py by just
substituting in the maths from the function to avoid calling the
function.

brandonhorsley committed:
Removed a FutureWarning I forgot to remove

ricardoV94 committed:
refactor helper function

ricardoV94 committed:
Tweak tests
